xs2a-adapter-service-api-adapter/src/main/java/de/adorsys/xs2a/adapter/http/ResponseHandlers.java

```diff
@@ -41,21 +54,51 @@ private ResponseHandlers() {
                 return jsonMapper.readValue(responseBody, klass);
             }
 
-            throw responseException(statusCode, new PushbackInputStream(responseBody), responseHeaders);
+            // will be invoked for not expected 2xx and 3xx statuses
+            throw responseException(statusCode, pushbackResponseBody, responseHeaders,
+                ResponseHandlers::buildErrorResponseFromString);
         };
     }
 
+    private static boolean isNotJson(PushbackInputStream responseBody) {
+        try {
+            int data = responseBody.read();
+            responseBody.unread(data);
+
+            if (data != -1) {
+                char firstChar = (char) data;
+                // There are only 2 possible char options here: '{' (JSON) and '<' (XML and HTML).
+                // ASCII code of '{' is 123 and ASCII code of '<' is 66,
+                // so the scope of the byte is enough for this purpose.
+                return firstChar != '{';
```

Comment:
Hm...it seems to me, that `JSON` could start with `[`. Am I wrong?

Reply:
It can start with [, event with ", or it can also be true/false/number.

https://jsonlint.com/?json=%7B%22hello%22%3A%22world%22%7D
https://jsonlint.com/?json=%5B%22hello%22%2C%20%22world%22%5D
https://jsonlint.com/?json=true
https://jsonlint.com/?json=1

Reply:
You can see more info here http://www.json.org/

Reply:
Thanks for the hints.
Anyway, in our case, I see no need to check these scenarios as the logic is the following:

- if `HTTP status is >= 400` AND `JSON starts with {`, it means that we can try to deserialize it into our model class for error messages (`ErrorResponse`);

- otherwise, we do know, that we couldn't proceed with deserialization and will get the Exception, so we prevent it by getting the error message as String.

`[`, `"`, `true`, `false`, `numbers` couldn't be deserialized into our model class.

Reply:
Maybe you could first check Content-Type and after that new logic without json Content-Type

Reply:
You mean smth like:

- `if HTTP status is >= 400`
        - `if (content type is JSON)` -> proceed the old way
        - `otherwise (content type is not provided OR is not JSON)`:
                 - `if response body doesn't start with {` -> proceed the new logic
                 - `otherwise` - proceed the old way

don't you?

Reply:
Something like this (probably should be more polished):

```java
if (contentType != null && !contentType.startsWith(APPLICATION_JSON)) {
    throw new NotAcceptableException(String.format(
                    "Content type %s is not acceptable, has to start with %s", contentType, APPLICATION_JSON));
}
```
modify to:
```java
if (contentType == null || !contentType.startsWith(APPLICATION_JSON)) {
    if (statusCode >= 400) {
        if (isNotJson(pushbackResponseBody)) {
            throw responseException(statusCode, pushbackResponseBody, responseHeaders,
                        ResponseHandlers::buildEmptyErrorResponse);
        }
        throw responseException(statusCode, pushbackResponseBody, responseHeaders,
                    ResponseHandlers::buildErrorResponseFromString);
    }
    throw new NotAcceptableException(String.format(
                    "Content type %s is not acceptable, has to start with %s", contentType, APPLICATION_JSON));
}
```

Reply:
It would be great if Spring (RestTemplate) could be used, with converters and well-configured RestTemplate most of this would work automatically.

Reply:
The problem with your solution is the following: if content type hasn't been provided, there will be an exception (even if the request body is correct). I'll think about how to handle this, thanks.

According to Spring:  it was our architecture design solution not to have Spring dependencies outside the REST modules.

Reply:
Ok, no problems

Reply:
Updated, thanks for your comments, folks